src/libraries/System.Net.NameResolution/src/System/Net/Dns.cs

```diff
@@ -665,5 +662,55 @@ private static bool LogFailure(ValueStopwatch stopwatch)
 
             return false;
         }
+
+        /// <summary>Mapping from key to current task in flight for that key.</summary>
+        private static readonly Dictionary<object, Task> s_tasks = new Dictionary<object, Task>();
+
+        /// <summary>Queue the function to be invoked asynchronously.</summary>
+        /// <remarks>
+        /// Since this is doing synchronous work on a thread pool thread, we want to limit how many threads end up being
+        /// blocked.  We could employ a semaphore to limit overall usage, but a common case is that DNS requests are made
+        /// for only a handful of endpoints, and a reasonable compromise is to ensure that requests for a given host are
+        /// serialized.  Once the data for that host is cached locally by the OS, the subsequent requests should all complete
+        /// very quickly, and if the head-of-line request is taking a long time due to the connection to the server, we won't
+        /// block lots of threads all getting data for that one host.  We also still want to issue the request to the OS, rather
+        /// than having all concurrent requests for the same host share the exact same task, so that any shuffling of the results
+        /// by the OS to enable round robin is still perceived.
+        /// </remarks>
+        private static Task<TResult> RunAsync<TResult>(Func<object, TResult> func, object key, CancellationToken cancellationToken)
+        {
+            Task<TResult>? task = null;
+
+            lock (s_tasks)
+            {
+                // Get the previous task for this key, if there is one
+                s_tasks.TryGetValue(key, out Task? prevTask);
+
+                // Schedule the work either as a continuation from the previous task (so that they're serialized)
+                // or immediately if there wasn't a previous task.
+                task = (prevTask ?? Task.CompletedTask).ContinueWith((prevTask, key) =>
```

Comment:
Nit: `prevTask` isn't needed in the continuation.
```suggestion
                task = (prevTask ?? Task.CompletedTask).ContinueWith((_, key) =>
```
?

Reply:
> prevTask isn't needed in the continuation

It's not, but I'm in the habit of doing that to shadow variables from the outer scope in case I do end up using it, thereby decreasing the size of the closure.